src/xen_exit_handler/src/xen_exit_handler.cpp

```diff
@@ -0,0 +1,266 @@
+
+void xen_exit_handler::handle_xen_wrmsr()
+{
+    auto val = 0ULL;
+
+    val |= ((m_state_save->rax & 0x00000000FFFFFFFF) << 0x00);
+    val |= ((m_state_save->rdx & 0x00000000FFFFFFFF) << 0x20);
+
+    bfdebug << vmcs::guest_cr3::get() << bfendl;
+    bfdebug << "hypervisor: " << std::hex << val << bfendl;
+    uintptr_t phys_addr = bfn::virt_to_phys_with_cr3(val, vmcs::guest_cr3::get());
+
+    auto imap = bfn::make_unique_map_x64<uintptr_t>(phys_addr);
+    uintptr_t *page = imap.get();
+
+    init_hypercall_page(page);
+    advance_rip();
+}
+
+void xen_exit_handler::complete_xen_vmcall(ret_type ret, vmcall_registers_t &regs)
+{
+    m_state_save->rax = regs.r00;
+    m_state_save->rdi = regs.r01;
+    m_state_save->rsi = regs.r02;
+    m_state_save->rdx = regs.r03;
+    m_state_save->r10 = regs.r04;
+    m_state_save->r08 = regs.r05;
+    m_state_save->r09 = regs.r06;
+    advance_rip();
+}
+
+void xen_exit_handler::init_start_info(vmcall_registers_t &regs)
+{
+    bfdebug << "RETRIEVED: " << std::hex << regs.r01 << bfendl;
+    auto imap = bfn::make_unique_map_x64<start_info_t>(regs.r01, vmcs::guest_cr3::get(),
+                                                       sizeof(start_info_t),
+                                                       vmcs::guest_ia32_pat::get());
+    start_info_t *start_info = imap.get();
+
+    strncpy(start_info->magic, "xen-TEST-TEST", 31);
+}
+
+
+
+
+
+void xen_exit_handler::init_shared_info(vmcall_registers_t &regs)
+{
+    auto imap = bfn::make_unique_map_x64<shared_info_t>(regs.r01, vmcs::guest_cr3::get(),
+                                                        sizeof(shared_info_t),
+                                                        vmcs::guest_ia32_pat::get());
+    shared_info = imap.get();
+    shared_info_addr = regs.r01;
+
+    shared_info->wc.version = 1;
+    shared_info->vcpu_info[0].time.version = 1;
+    tsc_khz = static_cast<unsigned int>(regs.r02);
+
+}
+
+
+
+void xen_exit_handler::update_fake_clock(vmcall_registers_t &regs)
+{
+    auto imap = bfn::make_unique_map_x64<shared_info_t>(regs.r01, vmcs::guest_cr3::get(),
+                                                        sizeof(shared_info_t),
+                                                        vmcs::guest_ia32_pat::get());
+    shared_info = imap.get();
+
+
+
+    uint64_t tsc_timestamp = shared_info->vcpu_info[0].time.tsc_timestamp;
+    uint32_t seconds = shared_info->wc.sec;
+    uint32_t nanoseconds = shared_info->wc.nsec;
+    uint64_t system_time = shared_info->vcpu_info[0].time.system_time;
+    uint64_t tsc = rdtsc();
+    uint64_t new_tsc = tsc;
+
+    tsc -= tsc_timestamp;
+    system_time += NANOSECONDS(tsc);
+    nanoseconds += static_cast<uint32_t>(system_time);
+    seconds += nanoseconds / 1000000000;
+    nanoseconds = nanoseconds % 1000000000;
+
+    //shared_info->vcpu_info[0].time.tsc_timestamp = new_tsc;
+    //shared_info->vcpu_info[0].time.system_time = system_time;
+    shared_info->wc.sec = seconds;
```

Comment:
i think during this, xen does something like setting the lowest bit during the update